### DISCUSSION TITLE
Tab bar icons do not load until the tab is activated

### DISCUSSION THREAD

**Discussion by givenergy-dexter** [expo/expo#26752] [Archived: Questions]
Hey all,

Just running into a weird issue with `Tabs` from `expo-router`. I've noticed that my icons are only being loaded once the tab has been activated.

| Home tab | Settings Tab |
|-----------|--------------|
| <img width="561" alt="Screenshot 2024-01-28 at 13 24 54" src="https://github.com/expo/expo/assets/150246412/c2a09df1-b569-4c1f-9bc7-1c0521264a49"> | <img width="561" alt="Screenshot 2024-01-28 at 13 25 07" src="https://github.com/expo/expo/assets/150246412/614fbf59-2a70-4430-9eaf-6c45cd921f09"> |

Here's my `Tabs.Screen` config that lives within the tab itself (`app/(app)/settings/index.js` not `app/(app)/_layout.js`):
```jsx
<Tabs.Screen
  options={{
    ...rest,
    tabBarIcon: ({ size, focused }) => (
      <Icon
        size={size}
        style={tw.style({
          'text-primary-faint': !focused,
          'text-primary': focused
        })}
      />
    )
  }}
/>
```

Is there a way for the icon to be loaded, without viewing the screen first, without having to manually define all the routes inside the `_layout.js` file?

Any constructive advice or suggestions are much appreciated.

**Comment by MarkSFrancis** [comment]
I've hit the same issue too. I'd like to define the screen options alongside the screen's definition, but obviously in the current structure, that would mean that route config (like the tab icon) wouldn't be fetched until the component loads. Because Expo doesn't fetch the tab's content (read: the code in the relevant `app/` file) until the tab is activated, you cannot set the options in that route. The same issue affects the [Drawer](https://docs.expo.dev/router/advanced/drawer/) component, and you can even see states where you'll have the`_layout` route options load first, and the dynamic route options load a few seconds afterwards if you don't use the SplashScreen properly when the app first loads (only removing the splash screen after the screen has set the navigation screen's options). 

In the future, I'd personally like to see some static export similar to NextJS' [getStaticProps](https://nextjs.org/docs/pages/building-your-application/data-fetching/get-static-props) function, which is exported outside of the default component code and therefore can be picked up at compilation and allow setting route options before the client loads the content. 

In the meantime, I think the recommendation should be to put all route config in the nearest `_layout`, and only use per-route config for options that are dynamic. Even then, you should probably still have some sort of "default" or "fallback" options in the `_layout` to display whilst the content is loading and the dynamic options are being configured

**Reply by MarkSFrancis** [reply]
You can work around some of these route-specific option quirks (like the screen's `title`) by using the [`SplashScreen`](https://docs.expo.dev/versions/latest/sdk/splash-screen/), but that doesn't help with navigation options like `tabBarIcon` that show before the content is loaded

|Not manually managing `SplashScreen`|Hiding `SpashScreen` when the route is loaded|
|-|-|
|https://github.com/expo/expo/assets/16414147/e63cbb22-1d56-4da5-bd5c-801c4d1b210e|https://github.com/expo/expo/assets/16414147/a25ab2cf-001f-494f-b86a-e039c6b683dd|

Notice the flash of "index" when the `SplashScreen` is not managed properly. This is because the app is trying to render the screen before the route-specific screen options have been loaded.

**Comment by vniehues** [comment]
@givenergy-dexter Did you ever find a solution for this? 
I think having to define the screen options in the nearest _layout file kinda defeats the purpose of having a file-based routing system. And I would really like to define the options in the routes themselves.
